### PR TITLE
Vocab

### PR DESCRIPTION
**Pull request by Iain-S**
Any tables marked as being vocab in the config file are downloaded during the `make-generators` step and saved as CSVs.

When the `create-vocab` command is used, the data is loaded from the CSVs.

### CODE REVIEW

File: sqlsynthgen/base.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+@dataclass
+class FileUploader:
+    """For uploading data files."""
+
+    table: Any
+
+    def load(self, connection: Any) -> None:
+        """Load the data from file."""
+        with Path(self.table.fullname + ".csv").open(
```
Comment (mhauru) [inline sqlsynthgen/base.py:18]: Might the table name be either an argument of load, or a field of the object?

File: sqlsynthgen/create.py [hunk 2]
```diff
@@ -39,8 +49,8 @@ def populate(
 ) -> None:
     """Populate a database schema with dummy data."""
 
-    for table, generator in zip(
-        tables, generators
+    for table, generator in reversed(
+        list(zip(reversed(tables), reversed(generators)))
```
Comment (mhauru) [inline sqlsynthgen/create.py:53]: Don't the reversals reverse each other?

File: sqlsynthgen/make.py [hunk 3]
```diff
@@ -136,13 +141,35 @@ def _add_generator_for_table(
     return content, new_class_name
 
 
+def _download_table(table: Any, engine: Any) -> None:
```
Comment (mhauru) [inline sqlsynthgen/make.py:144]: Do the commented out lines in this function still serve a purpose?

File: tests/test_base.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+"""Tests for the main module."""
```
Comment (mhauru) [inline tests/test_base.py:1]: Docstring seems out of date

File: tests/test_base.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+class VocabTests(TestCase):
+    """Module test case."""
+
+    def setUp(self) -> None:
+        """Pre-test setup."""
+
+        run_psql("providers.dump")
+
+        self.engine = create_engine(
+            "postgresql://postgres:password@localhost:5432/providers"
```
Comment (mhauru) [inline tests/test_base.py:36]: Maybe this sort of stuff could be somewhere central for all tests, like a fixture or some such?

File: tests/test_make.py [hunk 6]
```diff
@@ -1,25 +1,86 @@
...
 class MyTestCase(TestCase):
     """Module test case."""
 
+    def setUp(self) -> None:
+        """Pre-test setup."""
+
+        run_psql("providers.dump")
+
+        self.engine = create_engine(
+            "postgresql://postgres:password@localhost:5432/providers",
```
Comment (mhauru) [inline tests/test_make.py:42]: Same comment as above

File: tests/test_make.py [hunk 7]
```diff
@@ -1,25 +1,86 @@
...
 import yaml
+from sqlalchemy import Column, Integer, create_engine, insert
+from sqlalchemy.orm import declarative_base
+from sqlalchemy.pool import NullPool
 
 from sqlsynthgen import make
 from tests.examples import example_orm
+from tests.utils import run_psql
+
+# pylint: disable=protected-access
```
Comment (mhauru) [inline tests/test_make.py:16]: Does this need to apply to the whole file?

File: tests/test_make.py [hunk 8]
```diff
@@ -1,25 +1,86 @@
...
         with open(conf_path, "r", encoding="utf8") as f:
             config = yaml.safe_load(f)
 
-        actual = make.make_generators_from_tables(example_orm, config)
+        with patch("sqlsynthgen.make._download_table",) as mock_download, patch(
+            "sqlsynthgen.make.create_engine"
+        ) as mock_create_engine, patch("sqlsynthgen.make.get_settings"):
+            actual = make.make_generators_from_tables(example_orm, config)
+            mock_download.assert_called_once()
+            # self.assertEqual(, mock_download.call_args_list[0])
```
Comment (mhauru) [inline tests/test_make.py:68]: Unnecessary comment?

File: tests/test_make.py [hunk 9]
```diff
@@ -1,25 +1,86 @@
...
+            mock_create_engine.assert_called_once()
+
+        self.assertEqual(expected, actual)
+
+    def test__download_table(self) -> None:
```
Comment (mhauru) [inline tests/test_make.py:73]: Dunder in `test_____________download_table`

File: tests/test_providers.py [hunk 10]
```diff
@@ -47,10 +48,15 @@ def setUp(self) -> None:
         run_psql("providers.dump")
 
         self.engine = create_engine(
-            "postgresql://postgres:password@localhost:5432/providers"
+            "postgresql://postgres:password@localhost:5432/providers",
```
Comment (mhauru) [inline tests/test_providers.py:51]: As above. Also the `# pool=NullPool` below seems dead

File: tests/test_providers.py [hunk 11]
```diff
@@ -63,6 +69,7 @@ def test_column_value(self) -> None:
             key = provider.column_value(conn, Person, "sex")
 
         self.assertEqual("M", key)
+        # pass
```
Comment (mhauru) [inline tests/test_providers.py:72]: Unnecessary comment?

**Review by mhauru — COMMENTED** [review]
Looks good, just some small nits

File: sqlsynthgen/make.py (hunk 3 above)
Comment (Iain-S) [reply sqlsynthgen/make.py:144]: Nope, now removed.

**Review by mhauru — APPROVED** [review]

**Pull request merged**